File: cpp/demo/Freeze/library/LibraryI.cpp
// **********************************************************************
//
// Copyright (c) 2003-2007 ZeroC, Inc. All rights reserved.
//
// This copy of Ice is licensed to you under the terms described in the
// ICE_LICENSE file included in this distribution.
//
// **********************************************************************

#include <Freeze/Freeze.h>
#include <LibraryI.h>

using namespace std;
using namespace Demo;

BookI::BookI(const LibraryIPtr& library) :
    _library(library), _destroyed(false)
{
}

void
BookI::destroy(const Ice::Current&)
{
    IceUtil::Mutex::Lock lock(*this);
    if(_destroyed)
    {
        throw Ice::ObjectNotExistException(__FILE__, __LINE__);
    }
    _destroyed = true;

    try
    {
        _library->remove(description);
    }
    catch(const Freeze::DatabaseException& ex)
    {
        DatabaseException e;
        e.message = ex.message;
        throw e;
    }
}

Demo::BookDescription
BookI::getBookDescription(const Ice::Current&) const
{
    IceUtil::Mutex::Lock lock(*this);

    if(_destroyed)
    {
        throw Ice::ObjectNotExistException(__FILE__, __LINE__);
    }

    //
    // Returns a copy
    //
    return description;
}

string
BookI::getRenterName(const Ice::Current&) const
{
    IceUtil::Mutex::Lock lock(*this);

    if(_destroyed)
    {
        throw Ice::ObjectNotExistException(__FILE__, __LINE__);
    }

    if(rentalCustomerName.empty())
    {
        throw BookNotRentedException();
    }
    return rentalCustomerName;
}

void
BookI::rentBook(const string& name, const Ice::Current&)
{
    IceUtil::Mutex::Lock lock(*this);

    if(_destroyed)
    {
        throw Ice::ObjectNotExistException(__FILE__, __LINE__);
    }

    if(!rentalCustomerName.empty())
    {
        throw BookRentedException();
    }
    rentalCustomerName = name;
}

void
BookI::returnBook(const Ice::Current&)
{
    IceUtil::Mutex::Lock lock(*this);

    if(_destroyed)
    {
        throw Ice::ObjectNotExistException(__FILE__, __LINE__);
    }

    if(rentalCustomerName.empty())
    {
        throw BookNotRentedException();
    }
    rentalCustomerName.clear();;
}

// Needs to be extern for the Sun C++ 5.4 compiler
//
extern Ice::Identity
createBookIdentity(const string& isbn)
{
    //
    // Note that the identity category is important since the locator
    // was installed for the category 'book'.
    //
    Ice::Identity ident;
    ident.category = "book";
    ident.name = isbn;

    return ident;
}

class IsbnToBook
{
public:

    IsbnToBook(const Ice::ObjectAdapterPtr& adapter) :
        _adapter(adapter)
    {
    }

    BookPrx
    operator()(const string& isbn)
    {
        return BookPrx::uncheckedCast(_adapter->createProxy(createBookIdentity(isbn)));
    }

private:

    const Ice::ObjectAdapterPtr _adapter;
};

LibraryI::LibraryI(const Ice::CommunicatorPtr& communicator, 
                   const string& envName, const string& dbName,
                   const Freeze::EvictorPtr& evictor) :
    _evictor(evictor),
    _connection(Freeze::createConnection(communicator, envName)),
    _authors(_connection, dbName)
{
}

Demo::BookPrx
LibraryI::createBook(const Demo::BookDescription& description, const Ice::Current& c)
{
    IceUtil::Mutex::Lock lock(*this);

#if defined(__SUNPRO_CC)
    //
    // Strange CC bug (only when optimizing and raising BookExistsException)
    //
    BookPrx book;
    {
        book = IsbnToBook(c.adapter)(description.isbn);
    }
#else
    BookPrx book = IsbnToBook(c.adapter)(description.isbn);
#endif
    try
    {
        book->ice_ping();

        //
        // The book already exists.
        //
        throw BookExistsException();
    }
    catch(const Ice::ObjectNotExistException&)
    {
        //
        // Book doesn't exist, ignore the exception.
        //
    }

    BookPtr bookI = new BookI(this);
    bookI->description = description;

    //
    // Create a new Ice Object in the evictor, using the new identity
    // and the new Servant.
    //
    // This can throw EvictorDeactivatedException (which indicates an
    // internal error). The exception is currently ignored.
    //
    Ice::Identity ident = createBookIdentity(description.isbn);
    _evictor->add(bookI, ident);

    //
    // Add the isbn number to the authors map.
    //
    Ice::StringSeq isbnSeq;

    StringIsbnSeqDict::iterator p =  _authors.find(description.authors);
    if(p != _authors.end())
    {
        isbnSeq = p->second;
    }

    isbnSeq.push_back(description.isbn);
    _authors.put(StringIsbnSeqDict::value_type(description.authors, isbnSeq));

    return book;
}

Demo::BookPrx
LibraryI::findByIsbn(const string& isbn, const Ice::Current& c) const
{
    //
    // No locking is necessary since no internal mutable state is
    // accessed.
    //

    try
    {
        BookPrx book = IsbnToBook(c.adapter)(isbn);
        book->ice_ping();
        return book;
    }
    catch(const Ice::ObjectNotExistException&)
    {
        //
        // Book doesn't exist, return a null proxy.
        //
        return 0;
    }
}

Demo::BookPrxSeq
LibraryI::findByAuthors(const string& authors, const Ice::Current& c) const
{
    IceUtil::Mutex::Lock lock(*this);

    //
    // Lookup all books that match the given authors, and return them
    // to the caller.
    //
    StringIsbnSeqDict::const_iterator p =  _authors.find(authors);

    BookPrxSeq books;

    if(p != _authors.end())
    {
        books.reserve(p->second.size());
        transform(p->second.begin(), p->second.end(), back_inserter(books), IsbnToBook(c.adapter));
    }

    return books;
}

void
LibraryI::setEvictorSize(::Ice::Int size, const Ice::Current&)
{
    //
    // No synchronization necessary, _evictor is immutable.
    //
    _evictor->setSize(size);
}

void
LibraryI::shutdown(const Ice::Current& current)
{
    current.adapter->getCommunicator()->shutdown();
}

void
LibraryI::remove(const BookDescription& description)
{
    IceUtil::Mutex::Lock lock(*this);
    
    //
    // Note: no need to catch and retry on deadlock since all access to
    // _authors is serialized.
    //

    try
    {
        StringIsbnSeqDict::iterator p = _authors.find(description.authors);
        
        assert(p != _authors.end());

        //
        // Remove the isbn number from the sequence.
        //
        Ice::StringSeq isbnSeq  = p->second;
        isbnSeq.erase(remove_if(isbnSeq.begin(), isbnSeq.end(), bind2nd(equal_to<string>(), description.isbn)),
                         isbnSeq.end());
        
        if(isbnSeq.empty())
        {
            //
            // If there are no further associated isbn numbers then remove
            // the record.
            //
            _authors.erase(p);
        }
        else
        {
            //
            // Otherwise, write back the new record.
            //
            p.set(isbnSeq);
        }

        //
        // This can throw EvictorDeactivatedException (which indicates
        // an internal error). The exception is currently ignored.
        //
        _evictor->remove(createBookIdentity(description.isbn));
    }
    catch(const Freeze::DatabaseException& ex)
    {
        DatabaseException e;
        e.message = ex.message;
        throw e;
    }
}

File: java/demo/Freeze/library/LibraryI.java
// **********************************************************************
//
// Copyright (c) 2003-2007 ZeroC, Inc. All rights reserved.
//
// This copy of Ice is licensed to you under the terms described in the
// ICE_LICENSE file included in this distribution.
//
// **********************************************************************

import Demo.*;

class LibraryI extends _LibraryDisp
{
    public synchronized BookPrx
    createBook(BookDescription description, Ice.Current current)
        throws DatabaseException, BookExistsException
    {
        BookPrx book = isbnToBook(description.isbn, current.adapter);

        try
        {
            book.ice_ping();

            //
            // The book already exists.
            //
            throw new BookExistsException();
        }
        catch(Ice.ObjectNotExistException e)
        {
            //
            // Book doesn't exist, ignore the exception.
            //
        }

        //
        // Create a new book Servant.
        //
        BookI bookI = new BookI(this);
        bookI.description = description;

        Ice.Identity ident = createBookIdentity(description.isbn);

        //
        // Create a new Ice Object in the evictor, using the new
        // identity and the new Servant.
        //
        // This can throw EvictorDeactivatedException (which indicates
        // an internal error). The exception is currently ignored.
        //
        _evictor.add(bookI, ident);

        try
        {
            //
            // Add the isbn number to the authors map.
            //
            String[] isbnSeq = (String[])_authors.get(description.authors);
            int length = (isbnSeq == null) ? 0 : isbnSeq.length;
            String[] newIsbnSeq = new String[length+1];

            if(isbnSeq != null)
            {
                System.arraycopy(isbnSeq, 0, newIsbnSeq, 0, length);
            }
            newIsbnSeq[length] = description.isbn;

            _authors.fastPut(description.authors, newIsbnSeq);
        
            return book;
        }
        catch(Freeze.DatabaseException ex)
        {
            DatabaseException e = new DatabaseException();
            e.message = ex.message;
            throw e;
        }
    }

    //
    // No locking is necessary since no internal mutable state is
    // accessed.
    //
    public BookPrx
    findByIsbn(String isbn, Ice.Current current)
        throws DatabaseException
    {
        try
        {
            BookPrx book = isbnToBook(isbn, current.adapter);
            book.ice_ping();

            return book;
        }
        catch(Ice.ObjectNotExistException ex)
        {
            //
            // Book doesn't exist, return a null proxy.
            //
            return null;
        }
    }

    public synchronized BookPrx[]
    findByAuthors(String authors, Ice.Current current)
        throws DatabaseException
    {
        try
        {
            //
            // Lookup all books that match the given authors, and
            // return them to the caller.
            //
            String[] isbnSeq = (String[])_authors.get(authors);

            int length = (isbnSeq == null) ? 0 : isbnSeq.length;
            BookPrx[] books = new BookPrx[length];

            if(isbnSeq != null)
            {
                for(int i = 0; i < length; ++i)
                {
                    books[i] = isbnToBook(isbnSeq[i], current.adapter);
                }
            }

            return books;
        }
        catch(Freeze.DatabaseException ex)
        {
            DatabaseException e = new DatabaseException();
            e.message = ex.message;
            throw e;
        }
    }

    public void
    setEvictorSize(int size, Ice.Current current)
        throws DatabaseException
    {
        //
        // No synchronization necessary, _evictor is immutable.
        //
        _evictor.setSize(size);
    }

    public void
    shutdown(Ice.Current current)
    {
        current.adapter.getCommunicator().shutdown();
    }

    protected synchronized void
    remove(BookDescription description)
        throws DatabaseException
    {
        try
        {
            String[] isbnSeq = (String[])_authors.get(description.authors);

            assert isbnSeq != null;
            

            int i;
            for(i = 0; i < isbnSeq.length; ++i)
            {
                if(isbnSeq[i].equals(description.isbn))
                {
                    break;
                }
            }

            assert i < isbnSeq.length;

            if(isbnSeq.length == 1)
            {
                //
                // If there are no further associated isbn numbers then remove
                // the record.
                //
                _authors.fastRemove(description.authors);
            }
            else
            {
                //
                // Remove the isbn number from the sequence and write
                // back the new record.
                //
                String[] newIsbnSeq = new String[isbnSeq.length-1];
                System.arraycopy(isbnSeq, 0, newIsbnSeq, 0, i);
                if(i < isbnSeq.length - 1)
                {
                    System.arraycopy(isbnSeq, i+1, newIsbnSeq, i, isbnSeq.length - i - 1);
                }
            
                _authors.fastPut(description.authors, newIsbnSeq);
            }

            //
            // This can throw EvictorDeactivatedException (which
            // indicates an internal error). The exception is
            // currently ignored.
            //
            _evictor.remove(createBookIdentity(description.isbn));
        }
        catch(Freeze.DatabaseException ex)
        {
            DatabaseException e = new DatabaseException();
            e.message = ex.message;
            throw e;
        }
    }

    LibraryI(Ice.Communicator communicator, String envName, String dbName, Freeze.Evictor evictor)
    {
        _evictor = evictor;
        _connection = Freeze.Util.createConnection(communicator, envName);
        _authors = new StringIsbnSeqDict(_connection, dbName, true);
    }

    void
    close()
    {
        _authors.close();
        _connection.close();
    }

    private Ice.Identity
    createBookIdentity(String isbn)
    {
        //
        // Note that the identity category is important since the
        // locator was installed for the category 'book'.
        //
        Ice.Identity ident = new Ice.Identity();
        ident.category = "book";
        ident.name = isbn;

        return ident;
    }

    private BookPrx
    isbnToBook(String isbn, Ice.ObjectAdapter adapter)
    {
        return BookPrxHelper.uncheckedCast(adapter.createProxy(createBookIdentity(isbn)));
    }

    private Freeze.Evictor _evictor;
    private Freeze.Connection _connection;
    private StringIsbnSeqDict _authors;
}
